exonum/src/events/noise/wrappers/sodium_wrapper/handshake.rs

```diff
@@ -143,35 +192,46 @@ impl NoiseHandshake {
 }
 
 impl Handshake for NoiseHandshake {
-    fn listen<S>(self, stream: S) -> HandshakeResult<S>
+    type Result = Vec<u8>;
+
+    fn listen<S>(self, stream: S) -> HandshakeResult<S, Self::Result>
     where
         S: AsyncRead + AsyncWrite + 'static,
     {
         let peer_address = self.peer_address;
         let framed = self.read_handshake_msg(stream)
-            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+            .and_then(|(stream, handshake, _)| handshake.write_handshake_msg(stream, &[]))
             .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
-            .and_then(|(stream, handshake)| handshake.finalize(stream))
+            .and_then(|(stream, handshake, peer_list)| handshake.finalize(stream, peer_list))
             .map_err(move |e| {
                 e.context(format!("peer {} disconnected", peer_address))
                     .into()
             });
         Box::new(framed)
     }
 
-    fn send<S>(self, stream: S) -> HandshakeResult<S>
+    fn send<S>(self, stream: S) -> HandshakeResult<S, Self::Result>
     where
         S: AsyncRead + AsyncWrite + 'static,
     {
         let peer_address = self.peer_address;
-        let framed = self.write_handshake_msg(stream)
-            .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
-            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
-            .and_then(|(stream, handshake)| handshake.finalize(stream))
-            .map_err(move |e| {
-                e.context(format!("peer {} disconnected", peer_address))
-                    .into()
-            });
-        Box::new(framed)
+        match self.peer_list.clone() {
```

Comment:
We've discussed it elsewhere: There is no reason to use a message here, a `SocketAddr` (or, potentially, domain name + port) should suffice.